pkg/metrics/listener.go

```diff
@@ -29,24 +32,71 @@ var log = logf.RuntimeLog.WithName("metrics")
 // The metrics is on by default.
 var DefaultBindAddress = ":8080"
 
+// DefaultSecureBindAddress sets the default secure bind address for the metrics listener
+var DefaultSecureBindAddress = ":8843"
+
+// DefaultKeyPath is prefix path for TLS key
+var DefaultKeyPath = "tls.key"
+
+// DefualtCertPath is prefix path for TLS certificate
+var DefualtCertPath = "tls.crt"
+
 // NewListener creates a new TCP listener bound to the given address.
-func NewListener(addr string) (net.Listener, error) {
+func NewListener(addr, certDir string) (net.Listener, error) {
 	if addr == "" {
 		// If the metrics bind address is empty, default to ":8080"
-		addr = DefaultBindAddress
+		if certDir != "" {
+			addr = DefaultSecureBindAddress
+		} else {
+			addr = DefaultBindAddress
+		}
 	}
 
 	// Add a case to disable metrics altogether
 	if addr == "0" {
 		return nil, nil
 	}
 
+	var err error
+	if certDir != "" {
+		certPath := filepath.Join(certDir, DefualtCertPath)
+		certKey := filepath.Join(certDir, DefaultKeyPath)
+		certWatcher, err := certwatcher.New(certPath, certKey)
+		if err != nil {
+			return nil, err
+		}
+
+		go func() {
+			stop := make(chan struct{})
+			if err = certWatcher.Start(stop); err != nil {
+				log.Error(err, "certificate watcher error")
+				return
+			}
+		}()
+
+		cfg := &tls.Config{
+			NextProtos:     []string{"h2"},
+			GetCertificate: certWatcher.GetCertificate,
+		}
+
+		log.Info("metrics server is starting to listen", "addr", addr)
+		ln, err := tls.Listen("tcp", addr, cfg)
```

Comment:
Perhaps we could get a separate helper method that created the listener, then we can just call `getListener` and don't have to repeat the error handling? You could also create the listener and then wrap it using https://golang.org/pkg/crypto/tls/#NewListener which may make things a bit simpler? (Not sure)